### PR TITLE
Show file progress as a diagnostic while it occurs.

### PR DESCRIPTION
**Pull request by Julian**
Helps make sure someone using e.g. vim.diagnostic.goto_next
notices that the file isn't ready yet.

Closes: https://github.com/Julian/lean.nvim/issues/252

### CODE REVIEW

**Comment by Julian** [comment]
(The tests seem a bit flaky again locally, will investigate, but pinging the both of you in case you have an opinion on how well this works for you.)

File: lua/lean/_util.lua [hunk 1]
```diff
@@ -12,6 +12,20 @@ function M.tbl_repeat(value, times)
...
 
+--- Fetch the diagnostics for all Lean LSP clients from the current buffer.
+function M.lean_lsp_diagnostics(opts, bufnr)
+  bufnr = bufnr or 0
+  opts = opts or {}
+  local diagnostics = {}
+  for _, client in pairs(vim.lsp.buf_get_clients(bufnr or 0)) do
+    if client.name:match('^lean') then
+      opts.namespace = vim.lsp.diagnostic.get_namespace(client.id)
+      vim.list_extend(diagnostics, vim.diagnostic.get(0, opts))
```
Comment (rish987) [inline lua/lean/_util.lua:23]:
```suggestion
      vim.list_extend(diagnostics, vim.diagnostic.get(bufnr, opts))
```

Comment (Julian) [reply lua/lean/_util.lua:23]: Nice catch, I'd just noticed this myself, fixing, but first fighting with trying to get `vim.tbl_map` to do what I want.

Comment (Julian) [reply lua/lean/_util.lua]: (Spoiler: I gave up.)

File: lua/lean/_util.lua [hunk 2]
```diff
@@ -12,6 +12,19 @@ function M.tbl_repeat(value, times)
...
 end
 
+--- Fetch the diagnostics for all Lean LSP clients from the current buffer.
+function M.lean_lsp_diagnostics(opts, bufnr)
+  opts = opts or {}
+  local diagnostics = {}
+  for _, client in pairs(vim.lsp.buf_get_clients(bufnr)) do
+    if client.name:match('^lean') then
+      opts.namespace = vim.lsp.diagnostic.get_namespace(client.id)
+      vim.list_extend(diagnostics, vim.diagnostic.get(bufnr, opts))
```
Comment (rish987) [inline lua/lean/_util.lua:22]:
```suggestion
      vim.list_extend(diagnostics, vim.diagnostic.get(bufnr or 0, opts))
```
I think you need an explicit first parameter for the second to be considered.

File: lua/lean/_util.lua (hunk 1 above)
Comment (rish987) [reply lua/lean/_util.lua]: Huh, what exactly was the issue?

File: lua/lean/_util.lua (hunk 2 above)
Comment (Julian) [reply lua/lean/_util.lua:22]: Thanks!

**Comment by rish987** [comment]
LGTM as well, thank you! Yeah the tests in general need some work (I've been busy reading lately but will get back to #250 soon).

File: lua/lean/_util.lua (hunk 1 above)
Comment (Julian) [reply lua/lean/_util.lua]: I couldn't get the tests to pass after changing to:

```lua
function M.lean_lsp_diagnostics(opts, bufnr)
  bufnr = bufnr or 0
  opts = opts or {}
  return vim.tbl_flatten(vim.tbl_map(function(client)
    if not client.name:match('^lean') then return {} end
    opts.namespace = vim.lsp.diagnostic.get_namespace(client.id)
    return vim.diagnostic.get(bufnr, opts)
  end, vim.lsp.buf_get_clients(bufnr)))
end
```

and got fed up fighting with Lua.

**Pull request merged**